src/map/ai/controllers/player_controller.cpp

```diff
@@ -191,6 +192,7 @@ bool CPlayerController::WeaponSkill(uint16 targid, uint16 wsid)
                 PChar->pushPacket(new CMessageBasicPacket(PChar, PTarget, 0, 0, MSGBASIC_CANNOT_SEE));
                 return false;
             }
+            roeutils::event(ROE_WSKILL_USE, PChar, RoeDatagram("mob", (CMobEntity*)PTarget));
```

Comment:
I don't have much of a leg to stand on in this regard (see trust gambits), but this syntax is pretty funky.

Can we not just infer the type of the data in the datagram from the `ROE_WSKILL_USE` tag and cast it on the other side?

Also, the cast to MobEntity happens within event, so do you need to be doing it here?

Reply:
Hmm. I'm curious what seems funky about it? We "could" infer the type based on the event but one of the considerations I was keenly aware of with this was that there will undoubtedly be edge-cases later which may need passing different data in different circumstances, even for the same event. This was a core design principle, and why these Datagram structures worked well since I can pass Lua any arbitrary number I wish. *This* time I wanted to pass just a mob, but I can foresee a circumstance exists where there are multiple locations where an event trigger needs to be called, but different data / extra data should be passed in one but not the other. Lua can handle extra params if desired.

To stop these calls from being even messier/completely onerous, I use overloaded constructors to dynamically generate the Datagrams, which are built and assign type based on the type of the payload object given. This is why I explicitly cast here, and why I was always keen to explicitly specify the cast to what I wanted in the calls, especially with entities, as a wrong entity type here would mean a wrong entity type passed to Lua. It is correct that the additional cast in `event` is superflous, though, so I can certainly remove that (I believe it was leftover from before the overloaded constructors.)

Reply:
Oh, upon re-reading I felt I should point out that the first argument to the RoeDatagram constructor isn't repeating the type; it's a key-name you'd like the data to have when it's passed to Lua. Thus allowing passing multiple objects of the same type, just with different keys.